### PR TITLE
Sdl behaviour if VR not avaliable

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Created skeleton for supporting disabled interfaces on HMI side. 
Implemented supporting disabled VR inteface

### CODE REVIEW

**Comment by LuxoftAKutsan** [comment]
@pestOO @anosach-luxoft @okozlovlux @Anatoliy-Leshin please review

**Comment by LuxoftAKutsan** [comment]
@dtrunov please review

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 1]
```diff
@@ -194,6 +194,41 @@ bool CommandRequestImpl::CheckSyntax(const std::string& str,
...
+  return response;
+}
+
+bool CommandRequestImpl::ProcessHMIInterfacesAvailability(
+    const uint32_t hmi_correlation_id,
+    const hmi_apis::FunctionID::eType& function_id) {
+  HmiInterfaces& hmi_interfaces = application_manager_.hmi_interfaces();
+  HmiInterfaces::InterfaceName interface =
+      hmi_interfaces.GetInterfaceFromFunction(function_id);
+  DCHECK(interface != HmiInterfaces::HMI_INTERFACE_INVALID_ENUM);
```
Comment (AGaliuzov) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan how about release build? Seems like UB

Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @AGaliuzov no it is not UB. hmi_interfaces.GetInterfaceState is able to handle INVALID_ENUM

File: src/components/application_manager/include/application_manager/hmi_interfaces.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    HMI_INTERFACE_TTS,
+    HMI_INTERFACE_UI,
+    HMI_INTERFACE_Navigation,
+    HMI_INTERFACE_VehicleInfo,
+    HMI_INTERFACE_SDL
+  };
+
+  enum InterfaceState {
+    STATE_NOT_RESPONSE,
+    STATE_AVALIABLE,
```
Comment (anosach-luxoft) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @LuxoftAKutsan typo

File: src/components/application_manager/include/application_manager/hmi_interfaces_impl.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace application_manager {
+class HmiInterfacesImpl : public HmiInterfaces {
+ public:
+  HmiInterfacesImpl() {}
+  ~HmiInterfacesImpl() {}
+  InterfaceState GetInterfaceState(InterfaceName interface) const OVERRIDE;
+  InterfaceName GetInterfaceFromFunction(
+      hmi_apis::FunctionID::eType function) const OVERRIDE;
+  void SetIntefaceState(InterfaceName interface, InterfaceState state) OVERRIDE;
```
Comment (anosach-luxoft) [inline src/components/application_manager/include/application_manager/hmi_interfaces_impl.h]: @LuxoftAKutsan typo

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 4]
```diff
@@ -1427,6 +1432,8 @@ class ApplicationManagerImpl
    */
   resumption::ResumeCtrl resume_ctrl_;
 
+  HmiInterfacesImpl hmi_interfaces_;
```
Comment (anosach-luxoft) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LuxoftAKutsan why should AM hold this instance? Why not request controller?

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 5]
```diff
@@ -194,6 +194,41 @@ bool CommandRequestImpl::CheckSyntax(const std::string& str,
...
+  response[strings::params][strings::correlation_id] = hmi_correlation_id;
+  response[strings::params][strings::protocol_type] =
+      CommandImpl::hmi_protocol_type_;
+  response[strings::params][strings::protocol_version] =
+      CommandImpl::protocol_version_;
+  response[strings::params][strings::function_id] = function_id;
+  response[strings::params][hmi_response::code] =
+      hmi_apis::Common_Result::UNSUPPORTED_RESOURCE;
+  return response;
+}
```
Comment (anosach-luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan maybe to message helper?

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 6]
```diff
@@ -194,6 +194,41 @@ bool CommandRequestImpl::CheckSyntax(const std::string& str,
...
+  HmiInterfaces& hmi_interfaces = application_manager_.hmi_interfaces();
+  HmiInterfaces::InterfaceName interface =
+      hmi_interfaces.GetInterfaceFromFunction(function_id);
+  DCHECK(interface != HmiInterfaces::HMI_INTERFACE_INVALID_ENUM);
+  HmiInterfaces::InterfaceState state =
+      hmi_interfaces.GetInterfaceState(interface);
+  if (state == HmiInterfaces::STATE_NOT_AVALIABLE) {
+    event_engine::Event event(function_id);
+    event.set_smart_object(
+        CreateUnsupportedResourceResponse(function_id, hmi_correlation_id));
```
Comment (anosach-luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan I think it would be better to just return bool value from here and do not send any responses. And than send response explicitly.

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 7]
```diff
@@ -311,6 +311,15 @@ void RegisterAppInterfaceRequest::Run() {
   SendRegisterAppInterfaceResponseToMobile();
 }
 
+void FillVRRelatedFields(smart_objects::SmartObject& response_params,
```
Comment (anosach-luxoft) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan what is this? If this is an anonymous function please put it to anonymous namespace.

Comment (anosach-luxoft) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan btw why do we need it? I see this logic only once

**Comment by anosach-luxoft** [comment]
@LuxoftAKutsan please link JIRA ticket

File: src/components/application_manager/include/application_manager/application_manager.h [hunk 8]
```diff
@@ -47,6 +47,7 @@
 #include "application_manager/policies/policy_handler_interface.h"
 #include "application_manager/application_manager_settings.h"
 #include "application_manager/state_controller.h"
+#include "application_manager/hmi_interfaces.h"
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/application_manager.h]: @LuxoftAKutsan Maybe you need to use forward declaration enstead include?

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 9]
```diff
@@ -222,9 +251,17 @@ uint32_t CommandRequestImpl::SendHMIRequest(
...
+    LOG4CXX_DEBUG(logger_,
+                  "subscribe_on_event " << function_id << " "
+                                        << hmi_correlation_id);
+    subscribe_on_event(function_id, hmi_correlation_id);
+  }
+  if (ProcessHMIInterfacesAvailability(hmi_correlation_id, function_id)) {
+    if (!application_manager_.ManageHMICommand(result)) {
+      LOG4CXX_ERROR(logger_, "Unable to send request");
+      SendResponse(false, mobile_apis::Result::OUT_OF_MEMORY);
+    }
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan Is it correct send OUT_OF_MEMORY response if "ManageHMICommand" returns false?  implementation of "ManageHMICommand" contains a lot of casec when "ManageHMICommand" returns false.

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 10]
```diff
@@ -194,6 +194,41 @@ bool CommandRequestImpl::CheckSyntax(const std::string& str,
...
+bool CommandRequestImpl::ProcessHMIInterfacesAvailability(
+    const uint32_t hmi_correlation_id,
+    const hmi_apis::FunctionID::eType& function_id) {
+  HmiInterfaces& hmi_interfaces = application_manager_.hmi_interfaces();
+  HmiInterfaces::InterfaceName interface =
+      hmi_interfaces.GetInterfaceFromFunction(function_id);
+  DCHECK(interface != HmiInterfaces::HMI_INTERFACE_INVALID_ENUM);
+  HmiInterfaces::InterfaceState state =
+      hmi_interfaces.GetInterfaceState(interface);
+  if (state == HmiInterfaces::STATE_NOT_AVALIABLE) {
```
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan , please, replace them `HmiInterfaces::STATE_NOT_AVALIABLE  == state`

**Comment by dtrunov** [comment]
Reviewed

**Comment by pestOO** [comment]
@LuxoftAKutsan,
Do I need to duplicate comments from #741 ?

File: src/components/application_manager/include/application_manager/application_manager_impl.h (hunk 4 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @anosach-luxoft  And this components is not related to request controller 
Only few requests should access this instance and CommandImpl.

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 5 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @anosach-luxoft I don't wan't to share this function for others.

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 6 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @anosach-luxoft  I did it in first realisation. Callee code in that case will we too complicated.

File: src/components/application_manager/include/application_manager/hmi_interfaces_impl.h [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#ifndef HMI_INTERFACES_IMPL_H
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/hmi_interfaces_impl.h]: @LuxoftAKutsan Please write header guard according with google coding style https://google.github.io/styleguide/cppguide.html#The__define_Guard

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 9 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @dtrunov I reely don't know. It was moved from  https://github.com/smartdevicelink/sdl_core/pull/745/commits/6c54ba3c9434d20751ed3f3706fe29850f328380#diff-8096dfbe93a4cc16e2414dbd6ef6834dL227 
Probably if it fails, it means that we are unable ot alocate no more memory. Any way I don't think I should change this behaviour in scope of this PR

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc (hunk 7 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @anosach-luxoft I used this function to reduce complexity of SendRegisterAppInterfaceResponseToMobile method. I think that we should have few such functions for UI related fields, Navi related fields, etc ... 
I should not be in anonymous name space because it should belong to am::request  namespace

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 9 above)
Comment (dtrunov) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan Maybe create TODO for processing in future

File: src/components/application_manager/src/hmi_interfaces_impl.cc [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+void HmiInterfacesImpl::SetIntefaceState(HmiInterfaces::InterfaceName interface,
+                                         HmiInterfaces::InterfaceState state) {
+  interfaces_states_[interface] = state;
+}
+
+HmiInterfaces::InterfaceName HmiInterfacesImpl::GetInterfaceFromFunction(
+    hmi_apis::FunctionID::eType function) const {
+  // TODO(AKutsan): Generate map of functionid to inteface automaticaly from
+  // HMI_API.xml
```
Comment (dtrunov) [inline src/components/application_manager/src/hmi_interfaces_impl.cc]: @LuxoftAKutsan Should create task for processing this todo?

File: src/components/application_manager/include/application_manager/hmi_interfaces.h [hunk 13]
```diff
@@ -0,0 +1,35 @@
+#ifndef HMI_INTERFACES_H
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @LuxoftAKutsan Please write header guard according with google coding style https://google.github.io/styleguide/cppguide.html#The__define_Guard

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 2 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: a81c6c66c73eb0b2cb2606e20202e2e1e146ff89

File: src/components/application_manager/include/application_manager/hmi_interfaces_impl.h (hunk 3 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces_impl.h]: abf6544ddcf7f843a185554c238cdee1e83d4424

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 10 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: 710f0515634126b101ef7d3b7d054452e16b8d23

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 13 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @dtrunov 23b478c

File: src/components/application_manager/include/application_manager/hmi_interfaces_impl.h (hunk 11 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces_impl.h]: @dtrunov 23b478c

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 14]
```diff
@@ -194,6 +194,42 @@ bool CommandRequestImpl::CheckSyntax(const std::string& str,
...
 
+smart_objects::SmartObject CreateUnsupportedResourceResponse(
+    const hmi_apis::FunctionID::eType function_id,
+    const uint32_t hmi_correlation_id) {
+  smart_objects::SmartObject response;
+  smart_objects::SmartObject& params = response[strings::params];
+  params[strings::message_type] = MessageType::kResponse;
+  params[strings::correlation_id] = hmi_correlation_id;
+  params[strings::protocol_type] =
+      CommandImpl::hmi_protocol_type_;
```
Comment (pestOO) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan,
Do we really could not write it in one line?

File: src/components/application_manager/src/commands/hmi/vr_is_ready_response.cc [hunk 15]
```diff
@@ -49,9 +49,15 @@ void VRIsReadyResponse::Run() {
...
-
   HMICapabilities& hmi_capabilities = application_manager_.hmi_capabilities();
   hmi_capabilities.set_is_vr_cooperating(is_available);
+
+  const HmiInterfaces::InterfaceState inteface_state =
+      is_available ? HmiInterfaces::STATE_AVAILABLE
+                   : HmiInterfaces::STATE_NOT_AVAILABLE;
+  HmiInterfaces& hmi_interfaces = application_manager_.hmi_interfaces();
+  hmi_interfaces.SetInterfaceState(HmiInterfaces::HMI_INTERFACE_VR,
+                                  inteface_state);
```
Comment (pestOO) [inline src/components/application_manager/src/commands/hmi/vr_is_ready_response.cc]: @LuxoftAKutsan.
Could we avoid copy-pasta and move this logic to the base `Is_ready` response?
According to all `*IsReadyResponse::Runs` we can merge them all in a one.

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 16]
```diff
@@ -366,6 +374,10 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile() {
     result_code = mobile_apis::Result::WRONG_LANGUAGE;
   }
 
+  if (application_manager_.hmi_interfaces().GetInterfaceState(
```
Comment (pestOO) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan, 
`GetInterfaceState` return enum, not a `bool`. So? it's not correct to verify it this way.

Could we use something like `IsInterfaceAvalibe(Type):bool`?

File: src/components/application_manager/include/application_manager/hmi_interfaces.h [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_HMI_INTERFACES_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_HMI_INTERFACES_H_
+
+#include "interfaces/HMI_API.h"
+
+namespace application_manager {
+
+class HmiInterfaces {
```
Comment (pestOO) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @LuxoftAKutsan,
> [EZamakhov] ]It's not exactly Interfaces, but more like Interfaces provider/controller/manager.

> [Akutsan} could you please suggest better name
> HmiInterfacesProvider?
> HmiInterfacesController?
> HmiInterfacesManager?

@anosach-luxoft, @AGaliuzov, what do you think?

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc (hunk 16 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @pestOO  No, it han be in 3 different stated:
Avaliable
Not avaliable
NO_RESPONSE

File: src/components/application_manager/src/commands/hmi/vr_is_ready_response.cc (hunk 15 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/hmi/vr_is_ready_response.cc]: @pestOO I think It should be done in scope of implementing other related CRQ, to clarify is it going to be correct for each one.

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 14 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @pestOO clang_formater

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 18]
```diff
@@ -366,6 +374,10 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile() {
...
+          HmiInterfaces::HMI_INTERFACE_VR)) {
+    FillVRRelatedFields(response_params, hmi_capabilities);
+  }
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan Incorrect condition.

Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @dtrunov  b08c7a3

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc (hunk 16 above)
Comment (pestOO) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan,
In this case compare with expected enum value, please,

Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @pestOO fixed in adbea6701018675342ed1dc8fa1d89f348093cb1

File: src/components/application_manager/src/hmi_interfaces_impl.cc (hunk 12 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/hmi_interfaces_impl.cc]: f76479d

**Comment by LuxoftAKutsan** [comment]
@dtrunov @OHerasym @AGritsevich @okozlovlux @nk0leg @orhan-mehmedov 
Please review as soon as possible. This changes should b merged today.

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h [hunk 19]
```diff
@@ -231,6 +231,30 @@ class CreateInteractionChoiceSetRequest : public CommandRequestImpl {
    */
   bool IsWhiteSpaceExist(const smart_objects::SmartObject& choice_set);
 
+  /**
+   * @brief ProcessHmiError process received error from HMI.
+   * This function id not threasd safe. It should be protected with
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @LuxoftAKutsan typo threads

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h [hunk 20]
```diff
@@ -231,6 +231,30 @@ class CreateInteractionChoiceSetRequest : public CommandRequestImpl {
...
+   * vr_commands_lock_
+   * @param vr_result ERROR type
+   */
+  void ProcessHmiError(hmi_apis::Common_Result::eType vr_result);
+
+  /**
+   * @brief ProcessSuccesfulHMIResponse process succesfull response from HMI\
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @LuxoftAKutsan typo successful

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h [hunk 21]
```diff
@@ -231,6 +231,30 @@ class CreateInteractionChoiceSetRequest : public CommandRequestImpl {
...
+   * This function id not threasd safe. It should be protected with
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @LuxoftAKutsan typo

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h [hunk 22]
```diff
@@ -231,6 +231,30 @@ class CreateInteractionChoiceSetRequest : public CommandRequestImpl {
...
+   * vr_commands_lock_
+   * @param corr_id correlation id of received response
+   * @return true if resuest with corr_itd was sent on HMI, false otherwise
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @LuxoftAKutsan typo

File: src/components/application_manager/src/commands/mobile/delete_command_request.cc [hunk 23]
```diff
@@ -111,12 +111,85 @@ void DeleteCommandRequest::Run() {
   }
 }
 
+bool DeleteCommandRequest::CalculateResult() {
+  using namespace helpers;
+  // Invalid enum means that request didn't sen't
```
Comment (okozlovlux) [inline src/components/application_manager/src/commands/mobile/delete_command_request.cc]: @LuxoftAKutsan `send` typo

File: src/components/application_manager/include/application_manager/commands/mobile/register_app_interface_request.h [hunk 24]
```diff
@@ -85,7 +85,13 @@ class RegisterAppInterfaceRequest : public CommandRequestImpl {
    **/
   void SendRegisterAppInterfaceResponseToMobile();
 
- private:
+  void SendChangeRegistration(const hmi_apis::FunctionID::eType function_id,
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan please add comments for method

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 9 above)
Comment (dtrunov) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan please don't forget about TODO.

File: src/components/application_manager/include/application_manager/application_manager.h [hunk 25]
```diff
@@ -459,6 +460,8 @@ class ApplicationManager {
       uint32_t connection_key, const std::string& policy_app_id) const = 0;
 
   virtual resumption::ResumeCtrl& resume_controller() = 0;
+
+  virtual HmiInterfaces& hmi_interfaces() = 0;
```
Comment (AGritsevich) [inline src/components/application_manager/include/application_manager/application_manager.h]: @LuxoftAKutsan, doxygen here will be usefull, I think.

File: src/components/application_manager/src/commands/mobile/add_command_request.cc [hunk 26]
```diff
@@ -318,17 +320,24 @@ void AddCommandRequest::on_event(const event_engine::Event& event) {
       if (hmi_apis::Common_Result::SUCCESS != ui_result_) {
         (*message_)[strings::msg_params].erase(strings::menu_params);
       }
+      if (message[strings::msg_params].keyExists(strings::info)) {
+        info_ += message[strings::msg_params].asString();
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/add_command_request.cc]: @LuxoftAKutsan isn't correct string please change to "message[strings::msg_params][strings::info].asString()

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 17 above)
Comment (AGritsevich) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @anosach-luxoft, @AGaliuzov, @pestOO my vote for `HmiInterfacesProvider`.

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 27]
```diff
@@ -287,57 +288,75 @@ void CreateInteractionChoiceSetRequest::SendVRAddCommandRequests(
   LOG4CXX_DEBUG(logger_, "expected_chs_count_ = " << expected_chs_count_);
 }
 
+void CreateInteractionChoiceSetRequest::ProcessHmiError(hmi_apis::Common_Result::eType vr_result)
```
Comment (AGritsevich) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @LuxoftAKutsan , I think it's can be `const`.

**Comment by dtrunov** [comment]
@LuxoftAKutsan Reviewed

File: src/components/application_manager/include/application_manager/hmi_interfaces.h [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+    HMI_INTERFACE_VR,
+    HMI_INTERFACE_TTS,
+    HMI_INTERFACE_UI,
+    HMI_INTERFACE_Navigation,
+    HMI_INTERFACE_VehicleInfo,
+    HMI_INTERFACE_SDL
+  };
+
+  enum InterfaceState {
+    STATE_NOT_RESPONSE,
```
Comment (orhan-mehmedov) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @LuxoftAKutsan, What does it mean? No response, will not response, not yet received response? If the intention is to have not valid enum here - why not to name it state none or invalid?

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 29]
```diff
@@ -287,57 +288,75 @@ void CreateInteractionChoiceSetRequest::SendVRAddCommandRequests(
...
+    if (!error_from_hmi_) {
+      error_from_hmi_ = true;
+      std::string info = vr_result == hmi_apis::Common_Result::UNSUPPORTED_RESOURCE ?
+                  "VR is not supported by system" : "";
+      SendResponse(false,
+                   GetMobileResultCode(vr_result), info.c_str());
+    }
+}
+
+bool CreateInteractionChoiceSetRequest::ProcessSuccesfulHMIResponse(uint32_t corr_id)
```
Comment (AGritsevich) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @LuxoftAKutsan , I think input parameter can be `const`.

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 28 above)
Comment (pestOO) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @LuxoftAKutsan, 
Suggest to add doxygen with explanation each state.

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 30]
```diff
@@ -311,6 +311,102 @@ void RegisterAppInterfaceRequest::Run() {
...
+              hmi_response::num_custom_presets_available);
+
+      if (hmi_capabilities.display_capabilities()
+              ->getElement(hmi_response::image_capabilities)
+              .length() > 0) {
+        display_caps[hmi_response::graphic_supported] = true;
+      } else {
+        display_caps[hmi_response::graphic_supported] = false;
+      }
+
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan, isn't it better this way (instead if/else):
display_caps[hmi_response::graphic_supported] = 
0 < hmi_capabilities.display_capabilities()->getElement(hmi_response::image_capabilities).length();

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 31]
```diff
@@ -550,8 +580,42 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile() {
...
+      LOG4CXX_DEBUG(logger_, "Interface " << interface << "is not avaliable");
+  }
+}
+
+void RegisterAppInterfaceRequest::SendChangeRagistrationOnHMI(
+    ApplicationConstSharedPtr app) {
+  using namespace hmi_apis::FunctionID;
+  DCHECK(app);
+  if (!app) {
+    return;
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan, why not used DCHECK_OR_RETURN_VOID?

File: src/components/application_manager/src/commands/mobile/add_command_request.cc [hunk 32]
```diff
@@ -410,23 +441,43 @@ void AddCommandRequest::on_event(const event_engine::Event& event) {
...
   }
 
-  SendResponse(result, result_code, NULL, &(message[strings::msg_params]));
+  if (!BothSend() && is_vr_or_ui_unsupported) {
+      LOG4CXX_DEBUG(logger_, "!BothSend() && is_vr_or_ui_unsupported");
+      result = false;
+  }
+
+  if (is_vr_and_ui_unsupported) {
+      LOG4CXX_DEBUG(logger_, "is_vr_and_ui_unsupported");
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/commands/mobile/add_command_request.cc]: @LuxoftAKutsan, such log message is not very useful. Add more info and interpretation - like == true, or more words.

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 17 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @AGritsevich it does not provide any HMI interfaces. It just looks for it's states.

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 9 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: I won't add todo, because we have a lot of such places in code, I created jira task https://adc.luxoft.com/jira/browse/APPLINK-27022

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h (hunk 19 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @dtrunov 9a5e95b1f4934342764a11b2d20bb04b331fa799

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h (hunk 20 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @dtrunov 210b0456cab1fef22d0296c05637f9f242547479

File: src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h (hunk 21 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/mobile/create_interaction_choice_set_request.h]: @dtrunov 9a5e95b1f4934342764a11b2d20bb04b331fa799

File: src/components/application_manager/src/commands/mobile/delete_command_request.cc (hunk 23 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/delete_command_request.cc]: @okozlovlux 33a9ca24291bd9fb7db224841df4a93a443ddc7b

File: src/components/application_manager/include/application_manager/commands/mobile/register_app_interface_request.h (hunk 24 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/mobile/register_app_interface_request.h]: @dtrunov 8fd4a275a74e7f6ab8999b9cfe197d26a8439f50

File: src/components/application_manager/include/application_manager/application_manager.h (hunk 25 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/application_manager.h]: @AGritsevich 9e211d58ac9e33d8c8f4a2ad7bf3cc4fa48c4987

File: src/components/application_manager/src/commands/mobile/add_command_request.cc (hunk 26 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/add_command_request.cc]: @dtrunov a1e72b6fd158daf024c0fe7ae3c3136beb065c54

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc (hunk 27 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @AGritsevich 88ef098945dd0cb9ea555ef1616592f63c98cd8a

File: src/components/application_manager/include/application_manager/hmi_interfaces.h (hunk 28 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/hmi_interfaces.h]: @orhan-mehmedov @pestOO  87658341ba14360f00da6ef861689ce3277ab0ea

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc (hunk 29 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @AGritsevich 38383c69160b34d89f69b09fd8de11219bf2f7b2

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc (hunk 30 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @orhan-mehmedov f78e80f988ec82a77ee04b670c65a6155ff49904

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc (hunk 31 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: @orhan-mehmedov c94b1511a9e88aeb7659696c14f969b476a6e4c3

File: src/components/application_manager/src/commands/mobile/add_command_request.cc (hunk 32 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/add_command_request.cc]: @orhan-mehmedov 4ad1354953916195f63677d09cc06acb316f043e

**Comment by AGritsevich** [comment]
@LuxoftAKutsan, Reviewed.

**Comment by okozlovlux** [comment]
@LuxoftAKutsan , reviewed

**Comment by orhan-mehmedov** [comment]
@LuxoftAKutsan, Reviewed.

**Comment by OHerasym** [comment]
Reviewed.

**Comment by LuxoftAKutsan** [comment]
@anosach-luxoft please merge
